rkt/containers.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2014 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/coreos/rocket/pkg/lock"
+)
+
+type containerHandle struct {
+	uuid       string
+	lock       *lock.DirLock
+	isExited   bool
+	isGarbage  bool
+	isDeleting bool
+}
+
+type includeMask byte
+
+const (
+	includeContainersDir includeMask = 1 << iota
+	includeGarbageDir
+)
+
+// withContainerHandles iterates over the included containers calling function f for every container.
+func withContainerHandles(include includeMask, f func(*containerHandle) error) error {
+	ls, err := getContainers(include)
+	if err != nil {
+		return fmt.Errorf("failed to get containers: %v", err)
+	}
+
+	for _, uuid := range ls {
+		ch, err := getContainerHandle(uuid)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error getting container handle %q, ignoring\n", uuid)
+			continue
+		}
+		// TODO(vc): how to respond to f() errors, what's idiomatic for this pattern?
```

Comment:
I like the way the filepath.Walk handles errors:

See http://golang.org/src/path/filepath/path.go?s=9885:9951#L331